### ISSUE TITLE
Vim9: no error when accessing key from dictionary with non-string expression

### ISSUE THREAD

**Issue by lacygoill** [vim/vim#7771]
**Describe the bug**

In Vim9 script, no error is raised when accessing a key from a dictionary with a non-string expression.

**To Reproduce**

Run this shell command:

    vim -Nu NONE -S <(cat <<'EOF'
        vim9
        def Func()
            var d = {123: 0}
            echo d[123]
        enddef
        defcompile
    EOF
    )

No error is raised.

**Expected behavior**

An error is raised at compile time:

    E1012: Type mismatch; expected string but got number

**Environment**

 - Vim version: 8.2 Included patches: 1-2451
 - OS: Ubuntu 16.04.7 LTS
 - Terminal: xterm(363)

**Additional context**

When we try to add a key into a dictionary with a non-string expression, Vim doesn't automatically coerce it into a string.  Instead, it gives the error `E1012` at compile time:
```vim
vim9
def Func()
    var d = {[123]: 0}
enddef
Func()
```
    E1012: Type mismatch; expected string but got number

This seems inconsistent.  If it's disallowed to use a non-string expression when adding a key into a dictionary, then it should also be disallowed to use a non-string expression when accessing a key from a dictionary.

**Comment by lacygoill** [comment]
To be more clear, the fact that this line doesn't raise any error in the first snippet is expected:

    var d = {123: 0}

Because an automatic string coercion is expected in a literal dictionary.

What is not expected is that the next line doesn't raise any error:

    echo d[123]

Instead, we should have to write this:

    echo d['123']
           ^   ^

**Comment by lacygoill** [comment]
Sorry, the examples I gave were not clear enough.  These ones are better:
```vim
vim9
var d = {[1 + 2]: 0}
```
    E928: String required
```vim
vim9
var d = {3: 0}
echo d[1 + 2]
```
    no error

If `[1 + 2]` is disallowed in the first snippet, then it should also be disallowed in the second one; otherwise, the rule is confusing.

**Comment by lacygoill** [comment]
> otherwise, the rule is confusing.

Besides, it goes against type checking.  For example, what if we write an expression which sometimes wrongly evaluates to a bool?  In such a case, we probably want to be warned of this mistake; that's not what Vim does currently.  It coerces the bool into a string.
```vim
vim9
var d = {true: 0}
echo d[true]
```
    0

**Comment by lacygoill** [comment]
Or maybe people will find it too cumbersome to always use `string()` to cast the evaluation into a string.  In which case, the issue is not in this snippet:
```vim
vim9
var d = {3: 0}
echo d[1 + 2]
```
    no error

But in this one:
```vim
vim9
var d = {[1 + 2]: 0}
```
    E928: String required

Maybe Vim should automatically coerce `1 + 2` into a string, so that we don't have to use `string()`:
```vim
vim9
var d = {[string(1 + 2)]: 0}
```

**Comment by lacygoill** [comment]
Basically, it's a design choice; either, we give priority to type checking and `echo d[1 + 2]` should raise an error; or we give priority to usability, and `{[1 + 2]: 0}` should not raise an error.

**Comment by brammool** [comment]
Well, at least it needs to be consistent.  If we allow this for indexing a dictionary with a bool:
    var d = {[true]: 1, [false]: 2}
     echo d[true]
     echo d[false]
This currently doesn't work, but it could.  Similarly for a number:
    var days = {[1]: 'Monday', [2]: 'Tuesday'}
     echo days[1]
     echo days[2]

Currently this works:
  var d = {[string(true)]: 1, [string(false)]: 2}
  echo d[true]
Which is not consistent.

I do think that using a number or bool as the key is very useful, and requiring string() is not helping.
Thus I tend to think we need to adjust the dictionary literal to also do the conversion to string.
It's unlikely this will cause mistakes, right?

Converting something like a list to a string is already disallowed, I think that is good.

**Comment by lacygoill** [comment]
> I do think that using a number or bool as the key is very useful, and requiring string() is not helping.
> Thus I tend to think we need to adjust the dictionary literal to also do the conversion to string.
> It's unlikely this will cause mistakes, right?
>
> Converting something like a list to a string is already disallowed, I think that is good.

Yes, I think requiring `string()` is too cumbersome; at least for simple values, like numbers and booleans.
Vim already automatically coerces simple values into strings after `:echo`:
```vim
vim9
def Func()
    echo 1 .. 'one'
    echo true .. 'true'
enddef
Func()
```
So, it wouldn't be a brand new exception.  I don't think it will cause issues, as long as it's limited to simple values.

Edit: actually, the coercion is not a property of `:echo` but of the concatenation operator `..`.

**Comment by lacygoill** [comment]
Fixed by [8.2.2455](https://github.com/vim/vim/releases/tag/v8.2.2455).
